src/react/hooks/use-presence.ts

```diff
@@ -0,0 +1,151 @@
+    onStatusChange: params?.onConnectionStatusChange,
+  });
+  const { room, roomError, roomStatus } = useRoom({
+    onStatusChange: params?.onRoomStatusChange,
+  });
+
+  const [isPresent, setIsPresent] = useState(false);
+  const [error, setError] = useState<ErrorInfo | undefined>();
+
+  // store the roomStatus in a ref to ensure the correct value is used in the effect cleanup
+  const roomStatusAndConnectionStatusRef = useRef({ roomStatus, connectionStatus });
+
+  // we can't use the data param directly in a dependency array as it will cause an infinite loop
+  const dataRef = useRef(params);
+  useEffect(() => {
+    dataRef.current = params;
+  }, [params]);
+
+  useEffect(() => {
+    // Update the ref when roomStatus changes
+    roomStatusAndConnectionStatusRef.current = { roomStatus, connectionStatus };
+  }, [roomStatus, connectionStatus]);
+
+  // enter the room when the hook is mounted
+  useEffect(() => {
+    const canJoinPresence = roomStatus === RoomLifecycle.Attached && !INACTIVE_CONNECTION_STATES.has(connectionStatus);
+    const canLeavePresence =
+      roomStatusAndConnectionStatusRef.current.roomStatus === RoomLifecycle.Attached &&
+      !INACTIVE_CONNECTION_STATES.has(roomStatusAndConnectionStatusRef.current.connectionStatus);
+
+    // wait until the room is attached before attempting to enter, and ensure the connection is active
+    if (!canJoinPresence) return;
+    room.presence
+      .enter(dataRef.current?.enterWithData)
+      .then(() => {
+        setIsPresent(true);
+        setError(undefined);
+      })
+      .catch((error: unknown) => {
+        setError(error as ErrorInfo);
+      });
+
+    return () => {
+      // ensure we are still in an attached state before attempting to leave and the connection is active;
+      // a presence.leave call will produce an exception otherwise.
+      if (canLeavePresence) {
+        room.presence
+          .leave(dataRef.current?.leaveWithData)
+          .then(() => {
+            setIsPresent(false);
+            setError(undefined);
+          })
+          .catch((error: unknown) => {
+            setError(error as ErrorInfo);
+          });
+      }
+    };
+  }, [room, connectionStatus, roomStatus]);
+
+  // if provided, subscribes the user provided onDiscontinuity listener
+  useEffect(() => {
+    if (!params?.onDiscontinuity) return;
+    const { off } = room.presence.onDiscontinuity(params.onDiscontinuity);
+    return () => {
+      off();
+    };
+  }, [room, params]);
+
+  // memoize the methods to avoid re-renders and ensure the same instance is used
+  const update = useCallback(
+    (data?: PresenceData) =>
+      room.presence.update(data).then(() => {
+        setIsPresent(true);
+        setError(undefined);
+      }),
+    [room],
+  );
+
+  return {
+    connectionStatus,
+    connectionError,
+    roomStatus,
+    roomError,
+    update,
+    isPresent,
```

Comment:
With removal of `leave`, does `isPresent` make much sense anymore, as it will only be driven by normal lifecycle events?